File: src/php/YourSQL/YourSQL.php
<?php
/**
 * YourSQL - Conector MySQL Personalizado para PHP
 * Versión: 1.0.0
 * 
 * Proporciona una capa de abstracción sobre MySQLi con funcionalidades mejoradas
 */

namespace YourSQL;

use mysqli;
use mysqli_result;
use Exception;

class YourSQLConnection {
    private mysqli $connection;
    private string $host;
    private string $user;
    private string $password;
    private string $database;
    private int $port;
    private string $charset;
    private bool $connected = false;
    
    /**
     * Constructor - Inicializa los parámetros de conexión
     * 
     * @param string $host Dirección del servidor MySQL
     * @param string $user Usuario de la base de datos
     * @param string $password Contraseña del usuario
     * @param string $database Nombre de la base de datos
     * @param int $port Puerto de conexión (por defecto 3306)
     * @param string $charset Codificación de caracteres (por defecto utf8mb4)
     */
    public function __construct(
        string $host,
        string $user,
        string $password,
        string $database,
        int $port = 3306,
        string $charset = 'utf8mb4'
    ) {
        $this->host = $host;
        $this->user = $user;
        $this->password = $password;
        $this->database = $database;
        $this->port = $port;
        $this->charset = $charset;
    }
    
    /**
     * Establece la conexión con la base de datos
     * 
     * @return bool True si la conexión fue exitosa
     * @throws Exception Si la conexión falla
     */
    public function connect(): bool {
        try {
            $this->connection = new mysqli(
                $this->host,
                $this->user,
                $this->password,
                $this->database,
                $this->port
            );
            
            if ($this->connection->connect_error) {
                throw new Exception("Error de conexión: " . $this->connection->connect_error);
            }
            
            // Configurar charset
            $this->connection->set_charset($this->charset);
            $this->connected = true;
            
            return true;
        } catch (Exception $e) {
            echo "❌ Error al conectar con MySQL: " . $e->getMessage() . "\n";
            $this->connected = false;
            return false;
        }
    }
    
    /**
     * Cierra la conexión con la base de datos
     */
    public function disconnect(): void {
        if ($this->connected && $this->connection) {
            $this->connection->close();
            $this->connected = false;
        }
    }
    
    /**
     * Verifica si la conexión está activa
     * 
     * @return bool True si hay conexión activa
     */
    public function isConnected(): bool {
        return $this->connected && $this->connection && $this->connection->ping();
    }
    
    /**
     * Obtiene la conexión MySQLi
     * 
     * @return mysqli Conexión MySQLi
     */
    public function getConnection(): mysqli {
        if (!$this->isConnected()) {
            $this->connect();
        }
        return $this->connection;
    }
    
    /**
     * Ejecuta una consulta SQL
     * 
     * @param string $query Consulta SQL a ejecutar
     * @param array $params Parámetros para consultas preparadas (opcional)
     * @return mixed Resultado de la consulta
     */
    public function executeQuery(string $query, array $params = null) {
        if (!$this->isConnected()) {
            $this->connect();
        }
        
        try {
            // Si hay parámetros, usar consulta preparada
            if ($params !== null && count($params) > 0) {
                return $this->executePrepared($query, $params);
            }
            
            // Consulta simple
            $result = $this->connection->query($query);
            
            if ($result === false) {
                throw new Exception("Error en la consulta: " . $this->connection->error);
            }
            
            // Para consultas SELECT, devolver array de resultados
            if ($result instanceof mysqli_result) {
                $data = [];
                while ($row = $result->fetch_assoc()) {
                    $data[] = $row;
                }
                $result->free();
                return $data;
            }
            
            // Para INSERT, UPDATE, DELETE devolver filas afectadas
            return $this->connection->affected_rows;
            
        } catch (Exception $e) {
            echo "❌ Error ejecutando consulta: " . $e->getMessage() . "\n";
            echo "📋 Query: $query\n";
            return null;
        }
    }
    
    /**
     * Ejecuta una consulta preparada (prepared statement)
     * 
     * @param string $query Consulta con placeholders (?)
     * @param array $params Parámetros a vincular
     * @return mixed Resultado de la consulta
     */
    private function executePrepared(string $query, array $params) {
        $stmt = $this->connection->prepare($query);
        
        if ($stmt === false) {
            throw new Exception("Error preparando consulta: " . $this->connection->error);
        }
        
        // Detectar tipos de parámetros
        $types = '';
        foreach ($params as $param) {
            if (is_int($param)) {
                $types .= 'i';
            } elseif (is_float($param)) {
                $types .= 'd';
            } else {
                $types .= 's';
            }
        }
        
        // Vincular parámetros
        $stmt->bind_param($types, ...$params);
        
        // Ejecutar
        $stmt->execute();
        
        // Obtener resultados
        $result = $stmt->get_result();
        
        if ($result instanceof mysqli_result) {
            $data = [];
            while ($row = $result->fetch_assoc()) {
                $data[] = $row;
            }
            $stmt->close();
            return $data;
        }
        
        $affected = $stmt->affected_rows;
        $stmt->close();
        return $affected;
    }
    
    /**
     * Obtiene la lista de tablas en la base de datos
     * 
     * @return array Lista de nombres de tablas
     */
    public function getTables(): array {
        $result = $this->executeQuery("SHOW TABLES");
        if ($result) {
            return array_map(fn($row) => array_values($row)[0], $result);
        }
        return [];
    }
    
    /**
     * Obtiene información sobre las columnas de una tabla
     * 
     * @param string $tableName Nombre de la tabla
     * @return array Información de las columnas
     */
    public function getTableInfo(string $tableName): array {
        return $this->executeQuery("DESCRIBE $tableName") ?? [];
    }
    
    /**
     * Obtiene el ID del último registro insertado
     * 
     * @return int ID del último insert
     */
    public function getLastInsertId(): int {
        return $this->connection->insert_id;
    }
    
    /**
     * Escapa una cadena para prevenir SQL injection
     * 
     * @param string $string Cadena a escapar
     * @return string Cadena escapada
     */
    public function escapeString(string $string): string {
        if (!$this->isConnected()) {
            $this->connect();
        }
        return $this->connection->real_escape_string($string);
    }
    
    /**
     * Inicia una transacción
     */
    public function beginTransaction(): void {
        $this->connection->begin_transaction();
    }
    
    /**
     * Confirma una transacción
     */
    public function commit(): void {
        $this->connection->commit();
    }
    
    /**
     * Revierte una transacción
     */
    public function rollback(): void {
        $this->connection->rollback();
    }
    
    /**
     * Destructor - Cierra la conexión automáticamente
     */
    public function __destruct() {
        $this->disconnect();
    }
}


/**
 * Constructor de consultas SQL de forma programática
 */
class YourSQLQueryBuilder {
    private YourSQLConnection $connection;
    private string $select = '*';
    private string $from = '';
    private array $where = [];
    private array $orderBy = [];
    private ?int $limit = null;
    private ?int $offset = null;
    
    /**
     * Constructor
     * 
     * @param YourSQLConnection $connection Conexión a la base de datos
     */
    public function __construct(YourSQLConnection $connection) {
        $this->connection = $connection;
    }
    
    /**
     * Reinicia el estado del query builder
     * 
     * @return self
     */
    public function reset(): self {
        $this->select = '*';
        $this->from = '';
        $this->where = [];
        $this->orderBy = [];
        $this->limit = null;
        $this->offset = null;
        return $this;
    }
    
    /**
     * Define las columnas a seleccionar
     * 
     * @param string ...$columns Columnas a seleccionar
     * @return self
     */
    public function select(string ...$columns): self {
        $this->select = count($columns) > 0 ? implode(', ', $columns) : '*';
        return $this;
    }
    
    /**
     * Define la tabla principal
     * 
     * @param string $table Nombre de la tabla
     * @return self
     */
    public function from(string $table): self {
        $this->from = $table;
        return $this;
    }
    
    /**
     * Añade una condición WHERE
     * 
     * @param string $condition Condición WHERE
     * @return self
     */
    public function where(string $condition): self {
        $this->where[] = $condition;
        return $this;
    }
    
    /**
     * Añade ORDER BY
     * 
     * @param string $column Columna para ordenar
     * @param string $direction Dirección (ASC o DESC)
     * @return self
     */
    public function orderBy(string $column, string $direction = 'ASC'): self {
        $this->orderBy[] = "$column $direction";
        return $this;
    }
    
    /**
     * Añade LIMIT
     * 
     * @param int $limit Número de registros
     * @return self
     */
    public function limit(int $limit): self {
        $this->limit = $limit;
        return $this;
    }
    
    /**
     * Añade OFFSET
     * 
     * @param int $offset Desplazamiento
     * @return self
     */
    public function offset(int $offset): self {
        $this->offset = $offset;
        return $this;
    }
    
    /**
     * Construye la consulta SQL
     * 
     * @return string Consulta SQL construida
     */
    public function build(): string {
        $query = "SELECT {$this->select} FROM {$this->from}";
        
        if (count($this->where) > 0) {
            $query .= ' WHERE ' . implode(' AND ', $this->where);
        }
        
        if (count($this->orderBy) > 0) {
            $query .= ' ORDER BY ' . implode(', ', $this->orderBy);
        }
        
        if ($this->limit !== null) {
            $query .= " LIMIT {$this->limit}";
        }
        
        if ($this->offset !== null) {
            $query .= " OFFSET {$this->offset}";
        }
        
        return $query;
    }
    
    /**
     * Ejecuta la consulta construida
     * 
     * @return mixed Resultado de la consulta
     */
    public function execute() {
        $query = $this->build();
        $result = $this->connection->executeQuery($query);
        $this->reset();
        return $result;
    }
}

File: src/python/yoursql/yoursql.py
"""
YourSQL - Conector MySQL Personalizado
Versión: 1.0.0

Este módulo proporciona una capa de abstracción sobre mysql.connector
con funcionalidades adicionales y una interfaz más intuitiva.
"""

import mysql.connector
from mysql.connector import Error
from typing import Optional, Dict, Any, List, Tuple
import json


class YourSQLConnection:
    """
    Clase para gestionar conexiones a MySQL con funcionalidades mejoradas
    """
    
    def __init__(
        self,
        host: str,
        user: str,
        password: str,
        database: str,
        port: int = 3306,
        charset: str = 'utf8mb4',
        autocommit: bool = True
    ):
        """
        Inicializa una conexión a MySQL
        
        Args:
            host: Dirección del servidor MySQL
            user: Usuario de la base de datos
            password: Contraseña del usuario
            database: Nombre de la base de datos
            port: Puerto de conexión (por defecto 3306)
            charset: Codificación de caracteres (por defecto utf8mb4)
            autocommit: Auto-commit de transacciones (por defecto True)
        """
        self.config = {
            'host': host,
            'user': user,
            'password': password,
            'database': database,
            'port': port,
            'charset': charset,
            'autocommit': autocommit
        }
        
        self._connection: Optional[mysql.connector.MySQLConnection] = None
        self._cursor = None
        self._connected = False
        
    def connect(self) -> bool:
        """
        Establece la conexión con la base de datos
        
        Returns:
            bool: True si la conexión fue exitosa, False en caso contrario
        """
        try:
            self._connection = mysql.connector.connect(**self.config)
            self._connected = True
            return True
        except Error as e:
            print(f"❌ Error al conectar con MySQL: {e}")
            self._connected = False
            return False
    
    def disconnect(self):
        """Cierra la conexión con la base de datos"""
        try:
            if self._cursor:
                self._cursor.close()
        except:
            pass
        
        try:
            if self._connection:
                # Consumir resultados pendientes antes de cerrar
                if hasattr(self._connection, 'unread_result') and self._connection.unread_result:
                    self._connection.consume_results()
                    
                if self._connection.is_connected():
                    self._connection.close()
                    
            self._connected = False
        except:
            self._connected = False
            
    def is_connected(self) -> bool:
        """
        Verifica si la conexión está activa
        
        Returns:
            bool: True si hay conexión activa
        """
        return self._connected and self._connection and self._connection.is_connected()
    
    def get_cursor(self, dictionary: bool = True, buffered: bool = True):
        """
        Obtiene un cursor para ejecutar consultas
        
        Args:
            dictionary: Si True, devuelve resultados como diccionarios
            buffered: Si True, consume todos los resultados inmediatamente (recomendado)
            
        Returns:
            Cursor de MySQL
        """
        if not self.is_connected():
            self.connect()
            
        return self._connection.cursor(dictionary=dictionary, buffered=buffered)
    
    def execute_query(
        self,
        query: str,
        params: Optional[Tuple] = None,
        fetch_one: bool = False,
        fetch_all: bool = True
    ) -> Optional[Any]:
        """
        Ejecuta una consulta SQL
        
        Args:
            query: Consulta SQL a ejecutar
            params: Parámetros para consultas preparadas
            fetch_one: Si True, devuelve solo un resultado
            fetch_all: Si True, devuelve todos los resultados
            
        Returns:
            Resultados de la consulta o None
        """
        cursor = self.get_cursor(buffered=True)
        
        try:
            if params:
                cursor.execute(query, params)
            else:
                cursor.execute(query)
            
            # Detectar tipo de consulta
            query_upper = query.strip().upper()
            
            # Para consultas que devuelven resultados (SELECT, SHOW, DESCRIBE, etc.)
            if any(query_upper.startswith(cmd) for cmd in ['SELECT', 'SHOW', 'DESCRIBE', 'EXPLAIN']):
                if fetch_one:
                    result = cursor.fetchone()
                    return result
                elif fetch_all:
                    result = cursor.fetchall()
                    return result
                else:
                    return None
            
            # Para INSERT, UPDATE, DELETE
            self._connection.commit()
            return cursor.rowcount
            
        except Error as e:
            print(f"❌ Error ejecutando consulta: {e}")
            print(f"📋 Query: {query}")
            if params:
                print(f"📋 Params: {params}")
            return None
        finally:
            try:
                cursor.close()
            except:
                pass  # Ignorar errores al cerrar el cursor
    
    def execute_many(self, query: str, data: List[Tuple]) -> int:
        """
        Ejecuta una consulta múltiples veces con diferentes datos
        
        Args:
            query: Consulta SQL con placeholders
            data: Lista de tuplas con los datos
            
        Returns:
            int: Número de filas afectadas
        """
        cursor = self.get_cursor(dictionary=False, buffered=True)
        
        try:
            cursor.executemany(query, data)
            self._connection.commit()
            return cursor.rowcount
        except Error as e:
            print(f"❌ Error en execute_many: {e}")
            return 0
        finally:
            try:
                cursor.close()
            except:
                passcursor(dictionary=False)
        
        try:
            cursor.executemany(query, data)
            self._connection.commit()
            return cursor.rowcount
        except Error as e:
            print(f"❌ Error en execute_many: {e}")
            return 0
        finally:
            cursor.close()
    
    def get_tables(self) -> List[str]:
        """
        Obtiene la lista de tablas en la base de datos
        
        Returns:
            Lista con nombres de tablas
        """
        result = self.execute_query("SHOW TABLES", fetch_all=True)
        if result:
            return [list(row.values())[0] for row in result]
        return []
    
    def get_table_info(self, table_name: str) -> List[Dict]:
        """
        Obtiene información sobre las columnas de una tabla
        
        Args:
            table_name: Nombre de la tabla
            
        Returns:
            Lista con información de las columnas
        """
        return self.execute_query(f"DESCRIBE {table_name}", fetch_all=True)
    
    def __enter__(self):
        """Soporte para context manager"""
        self.connect()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Cierra la conexión al salir del contexto"""
        self.disconnect()
    
    def __del__(self):
        """Destructor - asegura que la conexión se cierre"""
        try:
            self.disconnect()
        except:
            pass  # Ignorar errores en el destructor


class YourSQLQueryBuilder:
    """
    Constructor de consultas SQL de forma programática
    """
    
    def __init__(self, connection: YourSQLConnection):
        self.connection = connection
        self.reset()
    
    def reset(self):
        """Reinicia el estado del query builder"""
        self._select = "*"
        self._from = ""
        self._where = []
        self._order_by = []
        self._limit = None
        self._offset = None
        return self
    
    def select(self, *columns: str):
        """Define las columnas a seleccionar"""
        self._select = ", ".join(columns) if columns else "*"
        return self
    
    def from_table(self, table: str):
        """Define la tabla principal"""
        self._from = table
        return self
    
    def where(self, condition: str):
        """Añade una condición WHERE"""
        self._where.append(condition)
        return self
    
    def order_by(self, column: str, direction: str = "ASC"):
        """Añade ORDER BY"""
        self._order_by.append(f"{column} {direction}")
        return self
    
    def limit(self, limit: int):
        """Añade LIMIT"""
        self._limit = limit
        return self
    
    def offset(self, offset: int):
        """Añade OFFSET"""
        self._offset = offset
        return self
    
    def build(self) -> str:
        """Construye la consulta SQL"""
        query = f"SELECT {self._select} FROM {self._from}"
        
        if self._where:
            query += " WHERE " + " AND ".join(self._where)
        
        if self._order_by:
            query += " ORDER BY " + ", ".join(self._order_by)
        
        if self._limit:
            query += f" LIMIT {self._limit}"
        
        if self._offset:
            query += f" OFFSET {self._offset}"
        
        return query
    
    def execute(self):
        """Ejecuta la consulta construida"""
        query = self.build()
        result = self.connection.execute_query(query)
        self.reset()
        return result
